### PR TITLE
Add withdraw from reward safe (update with DELEGATECALL)

### PR DESCRIPTION
**Pull request by tintinthong**
- this pr updates sdk removes the complex signing and uses alex implementaion of delegate call (it works)
- Note: I also made some updates to `executeTransaction` which now takes operation as an input. Before it just assumed everything is calls.

### CODE REVIEW

File: packages/cardpay-sdk/sdk/reward-manager/base.ts [hunk 1]
```diff
@@ -543,10 +477,10 @@ The owner of reward safe ${safeAddress} is ${rewardSafeOwner}, but the signer is
     }
     let fullSignature = await signRewardSafe(
       this.layer2Web3,
-      rewardManagerAddress,
-      0,
-      payload,
+      rewardSafeDelegateAddress,
       0,
+      withdrawPayload,
+      1,
```
Comment (habdelra) [inline packages/cardpay-sdk/sdk/reward-manager/base.ts:483]: let's use a constant here so it's clear that 1 stands for DELEGATECALL

File: packages/cardpay-sdk/sdk/reward-manager/base.ts [hunk 2]
```diff
@@ -558,10 +492,10 @@ The owner of reward safe ${safeAddress} is ${rewardSafeOwner}, but the signer is
 
     let eip1271Data = createEIP1271VerifyingData(
       this.layer2Web3,
-      rewardManagerAddress,
-      '0',
-      payload,
+      rewardSafeDelegateAddress,
       '0',
+      withdrawPayload,
+      '1',
```
Comment (habdelra) [inline packages/cardpay-sdk/sdk/reward-manager/base.ts:498]: let's use a constant here so it's clear that 1 stands for DELEGATECALL

File: packages/cardpay-sdk/sdk/reward-manager/base.ts [hunk 3]
```diff
@@ -572,8 +506,9 @@ The owner of reward safe ${safeAddress} is ${rewardSafeOwner}, but the signer is
     let gnosisTxn = await executeTransaction(
       this.layer2Web3,
       safeAddress,
-      rewardManagerAddress,
-      payload,
+      rewardSafeDelegateAddress,
+      withdrawPayload,
+      1,
```
Comment (habdelra) [inline packages/cardpay-sdk/sdk/reward-manager/base.ts:511]: let's use a constant here so it's clear that 1 stands for DELEGATECALL

File: packages/cardpay-sdk/sdk/reward-manager/base.ts [hunk 4]
```diff
@@ -436,102 +441,31 @@ The owner of reward safe ${safeAddress} is ${rewardSafeOwner}, but the signer is
...
     let estimate = await gasEstimate(
       this.layer2Web3,
       safeAddress,
-      rewardManagerAddress,
+      rewardSafeDelegateAddress,
       '0',
-      payloadNonIncrementingNonce,
-      0,
+      withdrawPayload,
+      1,
```
Comment (habdelra) [inline packages/cardpay-sdk/sdk/reward-manager/base.ts:457]: let's use a constant here so it's clear that 1 stands for DELEGATECALL

File: packages/cardpay-sdk/sdk/reward-pool/base.ts [hunk 5]
```diff
@@ -293,6 +293,7 @@ export default class RewardPool {
       safeAddress,
       tokenAddress,
       payload,
+      0,
```
Comment (habdelra) [inline packages/cardpay-sdk/sdk/reward-pool/base.ts:296]: let's use a constant here so it's clear that 0 stands for CALL

File: packages/cardpay-sdk/sdk/reward-pool/base.ts [hunk 6]
```diff
@@ -426,6 +427,7 @@ The reward program ${rewardProgramId} has balance equals ${fromWei(
       safeAddress,
       rewardPoolAddress,
       payload,
+      0,
```
Comment (habdelra) [inline packages/cardpay-sdk/sdk/reward-pool/base.ts:430]: let's use a constant here so it's clear that 0 stands for CALL

File: packages/cardpay-sdk/sdk/safes.ts [hunk 7]
```diff
@@ -360,6 +360,7 @@ export default class Safes implements ISafes {
       safeAddress,
       tokenAddress,
       payload,
+      0,
```
Comment (habdelra) [inline packages/cardpay-sdk/sdk/safes.ts:363]: let's use a constant here so it's clear that 0 stands for CALL

File: packages/cardpay-sdk/sdk/token-bridge-home-side.ts [hunk 8]
```diff
@@ -122,6 +122,7 @@ export default class TokenBridgeHomeSide implements ITokenBridgeHomeSide {
       safeAddress,
       tokenAddress,
       payload,
+      0,
```
Comment (habdelra) [inline packages/cardpay-sdk/sdk/token-bridge-home-side.ts:125]: let's use a constant here so it's clear that 0 stands for CALL

**Comment by habdelra** [comment]
@tintinthong did the SDK ever get updated with new contract addresses for the rewardpool and reward manager in sokol?

**Comment by tintinthong** [comment]
> @tintinthong did the SDK ever get updated with new contract addresses for the rewardpool and reward manager in sokol?

Yes. https://github.com/cardstack/cardstack/pull/2423/files

**Comment by tintinthong** [comment]
@habdelra I have added a numeric Enum to handle your suggestion. I so far have added the affected files. But there are a lot more places this can be used. I think I'll do this as a cleanup pr after both my prs are merged for rewards.

**Comment by habdelra** [comment]
@tintinthong I don't see your "enum" (I assume you mean const--since javascript doesn't understand the idea of an enum) in the codebase--can you point me to where you added it?

**Comment by tintinthong** [comment]
> @tintinthong I don't see your "enum" (I assume you mean const--since javascript doesn't understand the idea of an enum) in the codebase--can you point me to where you added it?

[Here](https://github.com/cardstack/cardstack/blob/1ea0859dc887846eb2f8117805d2c53d1d812aa8/packages/cardpay-sdk/sdk/utils/safe-utils.ts#L32) @habdelra

**Comment by habdelra** [comment]
Nice, I didn’t know that TS supported enum! Can you at least use that enum in this PR? And then in a subsequent PR we can clean up all the other uses of the operation magic constants in the SDK.

File: packages/cardpay-sdk/sdk/reward-manager/base.ts [hunk 9]
```diff
@@ -428,6 +429,11 @@ export default class RewardManager {
     let token = new this.layer2Web3.eth.Contract(ERC20ABI as AbiItem[], tokenAddress);
     let safeBalance = new BN(await token.methods.balanceOf(safeAddress).call());
 
+    let rewardSafeDelegateAddress = await getAddress('rewardSafeDelegate', this.layer2Web3);
```
Comment (alex-cardstack) [inline packages/cardpay-sdk/sdk/reward-manager/base.ts:432]: I think this is what we just discussed r.e. querying this from the reward manager contract

**Review by alex-cardstack — APPROVED** [review]
Looks good apart from the delegate implementation address thing that you have in progress

**Pull request merged**